test/functional/rpc_blockchain.py

```diff
@@ -505,6 +513,56 @@ def _test_stopatheight(self):
         self.start_node(0)
         assert_equal(self.nodes[0].getblockcount(), HEIGHT + 7)
 
+    def _test_waitfornewblock(self):
+        self.log.info("Test waitfornewblock")
+        node = self.nodes[0]
+        peer = node.add_p2p_connection(P2PInterface())
+
+        current_height = node.getblock(node.getbestblockhash())['height']
+        current_hash = node.getblock(node.getbestblockhash())['hash']
+
+        # Create a fork somewhere below our current height, invalidate and
+        # then reconsider the longest chain.
+        #
+        fork_height = current_height - 100 # choose something vaguely near our tip
+        fork_hash = node.getblockhash(fork_height)
+        fork_block = node.getblock(fork_hash)
+
+        b1 = self.solve_and_send_block(peer, int(fork_hash, 16), fork_height+1, fork_block['time'] + 1)
+        b2 = self.solve_and_send_block(peer, b1.sha256, fork_height+2, b1.nTime + 1)
+
+        node.invalidateblock(b2.hash)
+        node.reconsiderblock(b2.hash)
+
+        assert_equal(
+            node.waitfornewblock(timeout=2)['hash'],
+            current_hash)
```

Comment:
In commit "rpc: add test for waitforblock and waitfornewblock" (a18cb7adf192bae481c879fe865f1c159b2b8622)

I don't think I understand what this test is checking for. Maybe there could be a comment that describes what the test setup is doing what a failed result might look like. It seems like it is just creating a small fork and then invalidating the fork tip, so it shouldn't affect the most work tip or current chain. I would expect to see wait_until call after the invalidateblock call which verifies it has the intended effect. Also the 2 millisecond timeout seems small so if this is supposed to wait for something it doesn't seem like it waiting very long.

Same comments also apply to the  _test_waitforblock test below, and maybe it would be good to dedupe the test setup shared by these functions, moving it into a helper method they both could call.

Reply:
I don't fully understand the original test either. Will look into it later.